### ISSUE TITLE
Mutations performance characteristics

### ISSUE THREAD

**Issue by ibradwan** [ClickHouse/ClickHouse#29402]
Hi ClickHouse team,

I have some questions about the CH delete mutations:
1. Does the mutation processing time depend on the complexity of the mutation query?
2. Do update mutations perform significantly faster than the delete ones?
3. We noticed an increase in the query processing time during the mutations (like 2x). Is this expected? The metrics do not show a significant drop in CPU or memory during the mutation time?

Thank you!

**Comment by den-crane** [comment]
1. maybe.
2. Yes. If you have 400 columns in a table, update of a single column can be ~400 faster than delete.
3. yes. Mutations cause high disk i/o -> cache spoiling. Basically mutation reads all data from related partitions and uncompress all data / mutate / compress and write data back.

Mutations are admin operation implemented to satisfy GDPR. You should not use them on daily basis.

**Comment by ibradwan** [comment]
I see. So to give some context, our mutation query is just a plain filtering delete query, but the `WHERE` condition could get complex in terms of the number of conditions (but every condition is just equal, greater, ...etc). Also, the fields on which we filter could be non-primary or sort keys. Please check the sample query below for reference.

So in light of this, let me ask other questions regarding point one:
* In the mutation execution plan, I believe you apply the select query in order to know which parts & blocks to mutate. The effect of the query complexity on this part should be negligible compared to the total mutation time.
Now I suppose you check the filtering condition for every row you scan to see if it's going to be changed or not. The cost for this step should be almost equivalent to the step of running the query on the whole table (`Sum(applying a filter on one row) ~= applying the filter of the whole table`). If this is the case, then the query complexity should not be a big factor in the mutation time. This is probably a naive way of thinking about the mutations internals, but is there a reason that makes the query complexity contribution to the total mutation time significant? If yes, then do you have any recommendations for a query similar to the one below?
* Would it make a difference if we filter on primary/sort key fields only?
* Would it make a difference to add data-skipping indexes on non-primary/sort keys fields?

For debugability, is there a way to know if a mutation got stuck, or if it is still processing?

The problem we are facing is that under our privacy terms, we have to get rid of the data within 72 hours of the user's request to delete the data. We used the mutations for this purpose for a while, and it was taking less than 24 hrs. After the deletion requests to our system increased, we combined them into a single mutation. At some point, the mutation took 6 days and did not complete. This is why we are investigating the problem to see if it's a rare thing, or if it's going to be the case whenever the query gets slightly complex. FYI, our version is `20.10.7.4`.

A sample mutation query:
``` sql
ALTER TABLE
    clarity.page_data_local ON CLUSTER clickhouse_replica_set DELETE
WHERE
    (
        (
            ProjectId = 419791012508602
            AND IngestionTimestamp <= '2021-09-03 13:00:00'
        )
        OR (
            ProjectId = 123
            AND IngestionTimestamp <= '2019-11-07 13:00:00'
            AND (
                (
                    UserId = 4876
                    AND SessionId = 8469
                )
            )
        )
        OR (
            ProjectId = 123
            AND IngestionTimestamp <= '2019-11-07 13:00:00'
            AND (
                (
                    UserId = 4876
                    AND SessionId = 8469
                )
            )
        )
        OR (
            ProjectId = 123
            AND IngestionTimestamp <= '2019-11-07 13:00:00'
            AND (
                (
                    UserId = 4876
                    AND SessionId = 8469
                )
            )
        )
        OR (
            ProjectId = 632502520798305
            AND IngestionTimestamp <= '2021-09-23 11:56:06'
            AND BackendRegion = 'stageus2'
            AND (
                (
                    UserId = 1466459634
                    AND SessionId = 2626896147
                )
            )
        )
        OR (
            ProjectId = 632502520798305
            AND IngestionTimestamp <= '2021-09-23 12:07:30'
            AND BackendRegion = 'stageus2'
            AND (
                (
                    UserId = 1466459634
                    AND SessionId = 2626896147
                )
            )
        )
        OR (
            ProjectId = 632502520798305
            AND IngestionTimestamp <= '2021-09-08 15:11:07'
            AND BackendRegion = 'stageus2'
            AND (
                (
                    UserId = 1466459634
                    AND SessionId = 2626896147
                )
            )
        )
        OR (
            ProjectId = 632502520798305
            AND IngestionTimestamp <= '2021-09-09 13:52:34'
            AND BackendRegion = 'stageus2'
            AND (
                (
                    UserId = 1466459634
                    AND SessionId = 2626896147
                )
            )
        )
        OR (
            ProjectId = 632502520798305
            AND IngestionTimestamp <= '2021-09-09 13:52:13'
            AND BackendRegion = 'stageus2'
            AND (
                (
                    UserId = 1466459634
                    AND SessionId = 2626896147
                )
            )
        )
        OR (
            ProjectId = 632502520798305
            AND IngestionTimestamp <= '2021-09-09 13:53:10'
            AND BackendRegion = 'stageus2'
            AND (
                (
                    Timestamp >= '2021-09-01 00:00:00'
                    AND Timestamp <= '2021-09-07 00:00:00'
                )
            )
        )
        OR (
            ProjectId = 632502520798305
            AND IngestionTimestamp <= '2021-09-09 13:52:29'
            AND BackendRegion = 'stageus2'
            AND (
                (
                    Timestamp >= '2021-09-01 00:00:00'
                    AND Timestamp <= '2021-09-07 00:00:00'
                )
            )
        )
        OR (
            ProjectId = 632502520798305
            AND IngestionTimestamp <= '2021-09-09 13:55:07'
            AND BackendRegion = 'stageus2'
            AND (
                (
                    Timestamp >= '2021-09-01 00:00:00'
                    AND Timestamp <= '2021-09-07 00:00:00'
                )
            )
        )
        OR (
            ProjectId = 642159315712969
            AND IngestionTimestamp <= '2021-09-15 12:11:15'
            AND BackendRegion = 'stageus2'
            AND (
                (
                    Timestamp >= '2021-09-03 00:00:00'
                    AND Timestamp <= '2021-09-03 01:00:00'
                )
            )
        )
    );
```

**Comment by den-crane** [comment]
I have answers for your questions, I am not a developer. I am doing triage. Maybe other developers may answer.
Skip indexes are not used for sure.

20.10. is out of support.

>At some point, the mutation took 6 days and did not complete.

Can  you  provide `select * from system.mutations where not is_done format Vertical` 
How big is the parts in the column parts_to_do ?
Did you try to grep clickhouse server logs for mutation?

**Comment by ibradwan** [comment]
I noticed something in the `mutations` table. All the mutations get stuck on one part (i.e., `202108_138799_367074_106_370107`), and this part does not exist in the parts table!
I thought the problem is in the version, so I upgraded to `21.7.11.3` and still the problem was not fixed.

Here are the logs (this keeps repeating in a loop):
```
2021.09.29 14:16:07.539506 [ 12509 ] {} <Trace> clarity.page_data_local: Executing log entry to merge parts 202108_138799_367074_106_370107 to 202108_138799_367074_107_370107
2021.09.29 14:16:07.539578 [ 12509 ] {} <Warning> clarity.page_data_local: Part 202108_138799_367074_106_370107 is covered by 202108_138799_367074_106_370189 but should be merged into 202108_138799_367074_107_370107. This shouldn't happen often.
2021.09.29 14:16:07.539794 [ 12509 ] {} <Error> clarity.page_data_local: auto DB::StorageReplicatedMergeTree::processQueueEntry(ReplicatedMergeTreeQueue::SelectedEntryPtr)::(anonymous class)::operator()(DB::StorageReplicatedMergeTree::LogEntryPtr &) const: Code: 49, e.displayText() = DB::Exception: Part 202108_138799_367074_106_370107 is covered by 202108_138799_367074_106_370189 but should be merged into 202108_138799_367074_107_370107. This shouldn't happen often., Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0x8d3fe5a in /usr/bin/clickhouse
1. DB::Exception::Exception<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&>(int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x10147b53 in /usr/bin/clickhouse
2. DB::StorageReplicatedMergeTree::tryExecuteMerge(DB::ReplicatedMergeTreeLogEntry const&) @ 0x10136677 in /usr/bin/clickhouse
3. DB::StorageReplicatedMergeTree::executeLogEntry(DB::ReplicatedMergeTreeLogEntry&) @ 0x10126e14 in /usr/bin/clickhouse
4. ? @ 0x101dc1df in /usr/bin/clickhouse
5. DB::ReplicatedMergeTreeQueue::processEntry(std::__1::function<std::__1::shared_ptr<zkutil::ZooKeeper> ()>, std::__1::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&, std::__1::function<bool (std::__1::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&)>) @ 0x1056f16c in /usr/bin/clickhouse
6. DB::StorageReplicatedMergeTree::processQueueEntry(std::__1::shared_ptr<DB::ReplicatedMergeTreeQueue::SelectedEntry>) @ 0x1017089d in /usr/bin/clickhouse
7. ? @ 0x10323c97 in /usr/bin/clickhouse
8. ThreadPoolImpl<ThreadFromGlobalPool>::worker(std::__1::__list_iterator<ThreadFromGlobalPool, void*>) @ 0x8d83a38 in /usr/bin/clickhouse
9. ThreadFromGlobalPool::ThreadFromGlobalPool<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()>(void&&, void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()&&...)::'lambda'()::operator()() @ 0x8d855df in /usr/bin/clickhouse
10. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x8d80d1f in /usr/bin/clickhouse
11. ? @ 0x8d84603 in /usr/bin/clickhouse
12. start_thread @ 0x76db in /lib/x86_64-linux-gnu/libpthread-2.27.so
13. clone @ 0x12171f in /lib/x86_64-linux-gnu/libc-2.27.so
 (version 21.7.11.3 (official build))
```

**Comment by den-crane** [comment]
You can remove information about this part blocks from ZK manually.

Or install 21.8.8.29

ClickHouse release v21.8, 2021-08-12
Better handling of lost parts for ReplicatedMergeTree tables. Fixes rare inconsistencies in ReplicationQueue. Fixes #10368. #25820 (alesapin) Replace lost parts with empty parts instead of hacking replication queue

v21.9, 2021-09-09
Do not remove data on ReplicatedMergeTree table shutdown to avoid creating data to metadata inconsistency. #26716 (nvartolomei).
